src/impl/metadata/retriever/baseMetadataRetriever.ts

```diff
@@ -1,111 +1,255 @@
 
-  protected setCountQuery(countQuery: string, fetchSize: number) {
-    this.isLimitBasedQueryRetrieval = true;
-    this.countQuery = countQuery;
-    this.fetchSize = fetchSize;
-  }
-
   protected setQuery(query: string) {
     this.query = query;
-    if (this.isLimitBasedQueryRetrieval)
-      this.queryWithOffsetsAndLimit = this.query.concat(
-        ` LIMIT ${this.fetchSize} OFFSET 0`
-      );
+    this.countQuery = this.generateCountQuery();
   }
 
-  protected async getObjects(): Promise<T[]> {
-    let records: T[] = [];
-    const conn = this.org.getConnection();
-
-    // Not Limit and Offset, based so old method
-    if (!this.isLimitBasedQueryRetrieval) {
-      SFPowerkit.log(
-        `Method: isTooling :  ${this.tooling}, QUERY:  ${this.query}`,
-        LoggerLevel.TRACE
-      );
-
-      let result: QueryResult<T>;
-
-      // Query the org
-      if (this.tooling) {
-        result = await conn.tooling.query<T>(this.query);
-      } else {
-        result = await conn.query<T>(this.query);
-      }
-
-      records.push(...result.records);
-
-      while (!result.done) {
-        result = await this.queryMore(result.nextRecordsUrl);
-        records.push(...result.records);
-      }
-    } else {
-      SFPowerkit.log(
-        `Method: isToolingandLimitBasedQueryRetrieval : true, QUERY:  ${this.query}`,
-        LoggerLevel.TRACE
-      );
-
-      let retrievedRecordSize = 0;
-      let offset = 0;
-      this.totalSize = await this.getCount();
-
-      while (retrievedRecordSize < this.totalSize) {
-        SFPowerkit.log(
-          `To Retrieve Total Size:  ${this.totalSize},Retrieved Size:   ${retrievedRecordSize} , Current Offset: ${offset}`,
-          LoggerLevel.TRACE
-        );
-
-        let result: QueryResult<T>;
-        SFPowerkit.log(this.queryWithOffsetsAndLimit, LoggerLevel.TRACE);
-        result = await conn.tooling.query<T>(this.queryWithOffsetsAndLimit);
-        retrievedRecordSize += result.totalSize;
-
-        records.push(...result.records);
+  private generateCountQuery() {
+    let queryParts = this.query.toUpperCase().split("FROM");
+    let objectParts = queryParts[1].trim().split(" ");
+    let objectName = objectParts[0].trim();
+    this.objectName = objectName;
 
-        offset = offset + this.fetchSize;
-        this.queryWithOffsetsAndLimit = this.query.concat(
-          ` LIMIT ${this.fetchSize} OFFSET ${offset}`
-        );
-      }
-    }
+    let countQuery = `SELECT COUNT() FROM ${objectName}`;
 
-    return records;
+    return countQuery;
   }
 
-  private async queryMore(url: string): Promise<QueryResult<T>> {
+  protected async getObjects(): Promise<T[]> {
+    //let records: T[] = [];
     const conn = this.org.getConnection();
-    let result: QueryResult<T>;
+
     if (this.tooling) {
-      result = await conn.tooling.queryMore<T>(url);
+      return executeToolingQueryAsync(this.query, conn, this.objectName);
```

Comment:
@genoud  await?

Reply:
Yes, Not required as the return type is a promise. The caller can resolve it use await.